agent/consul/config_endpoint.go

```diff
@@ -252,22 +277,54 @@ func (c *ConfigEntry) ResolveServiceConfig(args *structs.ServiceConfigRequest, r
 				}
 			}
 
-			// Resolve the service definition by overlaying the service config onto the global
-			// proxy config.
-			definition := structs.ServiceDefinition{
-				Name: args.Name,
-			}
+			// Apply the proxy defaults to the sidecar's proxy config
 			if proxyConf != nil {
-				definition.Proxy = &structs.ConnectProxyConfig{
-					Config: proxyConf.Config,
+				if reply.Definition.Connect.SidecarService.Proxy == nil {
+					reply.Definition.Connect.SidecarService.Proxy = &structs.ConnectProxyConfig{
+						Config: proxyConf.Config,
+					}
+				} else if reply.Definition.Connect.SidecarService.Proxy.Config == nil {
+					reply.Definition.Connect.SidecarService.Proxy.Config = proxyConf.Config
+				} else {
+					for k, v := range proxyConf.Config {
+						reply.Definition.Connect.SidecarService.Proxy.Config[k] = v
+					}
```

Comment:
Note to self this is not quite right precedence. If you added `protocol = http` to the `proxy-defaults` it would overwrite the service-specific protocol we set above. We should probably move the block above down below this or otherwise change this to only merge unset.